### DISCUSSION TITLE
Auto completion in the middle of a sentence

### DISCUSSION THREAD

**Discussion by samuelgfeller** [community/community#77188] [Copilot Conversations] [answered]
## Select Topic Area

Product Feedback

## Body

When writing documentation, copilot is a great help with sentence autocompletion, but it seems to not work when there is text after where my cursor is.

### Example
If I have the following text and want a completion between the 2 parts of the sentence.

`Routes that need authentication can  `*[cursor here]*` adding a middleware to the route definition.`

If I start to write the following I get a pertinent completion suggestion:
`Routes that need authentication can ` _be protected by adding a middleware to the route definition._

Similarly, in the code when there are characters after the cursor, Copilot refuses to suggest anything **even when the executing the shortcut "Show Completions"**, "Show Previous Completions" and "Show Next Completions"

`$loginUrl = $this->routeParser->urlFor('`*[cursor here and "Show Completions" shortcut executed]*`', [], $queryParams);`   

And it knows what is correct when I start writing just the variable name:  
`$loginUrl = ` *$this->routeParser->urlFor('login-page', [], $queryParams);*

### Feedback
I wish that when I have the end of the sentence it still suggests me a completion taking into account the next part of the text, so in the example the missing part was `be protected by`. That is quite a big deal for me as it often happens that I have to break the line before the second part of the sentence to get the suggestion and then try to stick it back together.

Is that a known limitation that it doesn't work when in the middle of the text? I've also observed this with code. 

When trying to force a suggestion with the shortcut, the copilot icon is not even replaced by the loading icon like it does when it's “thinking”

**Comment by CheetahCodes21** [comment, marked as answer]
Yes, you're correct that Copilot currently has limitations when suggesting completions in the middle of text. This is because Copilot relies on contextual information to predict the next word or phrase, and it can be more challenging to provide accurate suggestions when there is existing text after the cursor.

The development team is aware of this limitation and is working on improvements. In the meantime, there are a few workarounds you can try:

1. **Break the line before the second part of the sentence:** As you mentioned, this can be a bit disruptive to your writing flow, but it can help Copilot provide more accurate suggestions.

2. **Delete the text after the cursor:** This will allow Copilot to analyze the complete context and provide more relevant suggestions.

3. **Use the shortcuts to cycle through suggestions:** Even though Copilot may not initially provide suggestions when you have text after the cursor, you can still use the shortcuts to cycle through potential completions. This can be helpful in narrowing down the options and finding the right fit.

4. **Provide more context:** Sometimes, providing more context in the surrounding text can help Copilot make better predictions. This could involve adding additional details, clarifying the purpose of the sentence, or providing more examples.

As Copilot continues to develop, we can expect to see improvements in its ability to provide suggestions in the middle of text. However, until then, these workarounds can help you get the most out of Copilot in the meantime.

**Reply by samuelgfeller** [reply]
Looking forward to this! I hope it will be able to include the next part of the text as part of its contextual information soon, that'd be fire!

**Reply by samuelgfeller** [reply]
I don't know if you can bring that up with relevant to the team or if they're already specifically working on this, but there are instances where I have the end of a sentence or code line and am looking for the beginning. It would be so amazing if it was able to generate suggestions backwards for the beginning of the sentence.

**Comment by hablutzel1** [comment]
Windsurf already implements this. Is GitHub Copilot bringing this too?
